[#40, #41, #42, #43 ]게시글 작성, 조회, 삭제,수정, 추천,비추천 기능 구현

donsonioc2010 opened the pull request:
## 기능 정리
1. 조회
    - Parameter로 받은 articleId를 기반으로 게시글을 조회한다.

2. 작성

3. 수정
4. 삭제
    - Parameter로 받은 ArticleId의 게시물 삭제 기능

5. 추천, 비추천
    - 게시물에 대한 추천
    - 게시물에 대한 비추천
    - 해당 게시물에 대한 추천, 비추천 갯수 조회

donsonioc2010 committed:
feat : No Handler Found Exception과 관련한 셋팅 추가

donsonioc2010 committed:
feat : 게시물 조회 기능 구현

donsonioc2010 committed:
refactor : 미사용 Import 제거

donsonioc2010 committed:
refactor : 권한 반환기능 오타 수정

donsonioc2010 committed:
feat : Article이 존재하지 않는 경우 Exception추가

donsonioc2010 committed:
test : 게시물 조회와 관련한 테스트케이스 추가

donsonioc2010 committed:
feat : timestamp 타입 변경

donsonioc2010 committed:
feat : Session의 로그인 여부 Check 확인 AOP 구현

donsonioc2010 committed:
feat : 로그인이 안된경우 UNAUTH코드에 대한 Return추가 및 AOP 설정 추가

donsonioc2010 committed:
feat : SessionUtils 현재 사용자의 UserRole추출기능 구현

donsonioc2010 committed:
feat : 게시글의 작성 및 수정기 구현

donsonioc2010 committed:
refactor : 테스트용 코드 제거

donsonioc2010 committed:
refactor : 조건문에 대해서 이해하기 쉽도록 명칭 수정

donsonioc2010 committed:
docs : 주석수정

donsonioc2010 committed:
feat : UserDto TestCase작성시 사용을 위한 Builder추가

donsonioc2010 committed:
test : UpsertArticle에 대한 TestCase추가

donsonioc2010 committed:
feat : ResponseDto의 Ok의 결과에 대해서 제작하는 메소드 추가 및 Enum의 불필요 메소드 삭제

donsonioc2010 committed:
feat : 게시물 삭제기능 구현

donsonioc2010 committed:
test : 게시물 삭제에 대한 테스트 케이스

donsonioc2010 committed:
feat : 게시물의 추천, 비추천의 추가,삭제,조회 기능 구현

donsonioc2010 committed:
feat : 게시물 추천, 비추천 기능 제작으로 인하여 기능 수정

- 게시물의 삭제를 진행할 경우 이전에 해당 게시물의 모든 추천과 비추천의 내역을 삭제한다.

donsonioc2010 committed:
test : 게시물의 추천, 비추천에 대한 테스트케이스 구현

- 비추천테스트케이스의 경우에는 같은 추천 메소드와 같은 메소드를 사용중인 상황이라 따로 구현하지 않음.

f-lab-kyle commented on src/main/java/com/nooblol/board/controller/ArticleController.java:
```diff
@@ -0,0 +1,43 @@
+      @PathVariable int articleId, HttpSession session
+  ) {
+    ArticleDto article = articleService.getArticleInfo(
+        articleId, SessionUtils.getSessionUserId(session)
```
SessionUtils.getSessionUserId(session) 이게 null이면 실행되지 않아야 할 녀석들도 실행되는거 아닐까요~?

donsonioc2010 replied:
해당 메소드에서 userId를 주고 있는 이유는 해당 사용자의 권한을 확인하기 위해서 전달해주는 부분이라 존재 유무를 알기 위함이어서 일부로 null을 제공하고 있습니다.

f-lab-kyle commented on src/main/java/com/nooblol/global/utils/UserUtils.java:
```diff
@@ -0,0 +1,29 @@
+   * @throws NoSuchAlgorithmException
+   * @throws UnsupportedEncodingException
+   */
+  public static String stringChangeToSha512(String value)
+      throws NoSuchAlgorithmException, UnsupportedEncodingException {
+    if (StringUtils.isBlank(value)) {
+      return null;
+    }
+    MessageDigest md = MessageDigest.getInstance("SHA-512");
+    md.update(value.getBytes(StandardCharsets.UTF_8));
+    return Base64.getEncoder().encodeToString(md.digest());
+  }
```
요건 암호화 유틸 이라 명칭하면 좋겠네요.

f-lab-kyle commented on src/main/java/com/nooblol/user/controller/AdminController.java:
```diff
@@ -0,0 +1,61 @@
+@RequiredArgsConstructor
+public class AdminController {
+
+  private final Logger log = LoggerFactory.getLogger(getClass());
```
`@Slf4j`

f-lab-kyle commented on src/main/java/com/nooblol/user/service/impl/AdminServiceImpl.java:
```diff
@@ -0,0 +1,178 @@
+      userSignUpRequestDto.setAdminUserRole();
+
+      ResponseDto returnDto = ResponseEnum.OK.getResponse();
+      if (userSignUpMapper.insertSignUpUser(userSignUpRequestDto) == 0) {
+        returnDto.setResult(false);
+      } else {
+        returnDto.setResult(true);
+      }
```
```suggestion
      returnDto.setResult(userSignUpMapper.insertSignUpUser(userSignUpRequestDto) != 0);
```

donsonioc2010 committed:
Merge pull request #52 from f-lab-edu/feature/#41

[#41 ] 게시글 작성 및 수정 기능 구현

donsonioc2010 committed:
Merge pull request #53 from f-lab-edu/feature/#42

[#42 ] 게시물 삭제 기능 구현

donsonioc2010 committed:
Merge pull request #54 from f-lab-edu/feature/#43

[#43 ] 게시글의 추천, 비추천 기능 구현

donsonioc2010 committed:
Merge Develop, feature/#41, 42, 43 Branch

donsonioc2010 committed:
refactor : Merge이후 수정안된 부분 변경

donsonioc2010 committed:
refactor : 하단 세부내역..

- Controller Mapping에 대해 기능별로 세부적으로 수정
- Mapper주석 불필요하다 판단하여 삭제
- 사용자의 관리자 여부 확인에 대한 검증 메소드 Enum으로 변경
- 조건문 삼항연산자 삭제 및 바로 결과값 Return하도록 수정

donsonioc2010 committed:
refactor : Dto final선언에 따른 Controller수정 및 Enum 로직오류 수정

donsonioc2010 commented on src/main/java/com/nooblol/board/controller/ArticleController.java:
```diff
@@ -0,0 +1,153 @@
+   * @param articleId
+   * @return
+   */
+  @GetMapping("/status/{articleId}")
```
여기서부터 아래의 메소드들은 추천이 대한 기능들인데요,
아래의 메소드들을 새로운 컨트롤러로 따로 제작하는게 좋았을까요? 같이 넣으면서 판단이 잘 서질 않았던거 같습니다.

f-lab-kyle replied:
기능의 구분이 명확하다면 나누는것도 좋아보입니다.

donsonioc2010 commented on src/main/java/com/nooblol/board/controller/ArticleController.java:
```diff
@@ -0,0 +1,153 @@
+      @PathVariable int articleId, HttpSession session
+  ) {
+    ArticleDto article = articleService.getArticleInfo(
+        articleId, SessionUtils.getSessionUserId(session)
```
해당 메소드에서는 사용자 Session에 대해서 로그인 유무를 체크하지 않았지만,
세션의 사용자 정보가 Null이어도 되도록 처리를 고의적으로 진행하였습니다.

사유는 Null로 들어올 경우에는 로그인이 되어있지 않다는 뜻이어서, 해당 Article에 대해서 Guest로써 게시물에 대해서 '볼 수만 있다' 는 형식으로 분류를 하기 위해서 처리했습니다.

f-lab-kyle replied:
👍

donsonioc2010 committed:
refactor

f-lab-kyle commented on src/main/java/com/nooblol/board/dto/ArticleStatusDto.java:
```diff
@@ -0,0 +1,30 @@
+  private String userId;
+
+  //True : 추천 , False : 비추천
+  private boolean type;
```
enum으로 관리해도 좋을것 같네요.

f-lab-kyle commented on src/main/java/com/nooblol/board/dto/ArticleStatusDto.java:
```diff
@@ -0,0 +1,30 @@
+  //True : 추천 , False : 비추천
+  private boolean type;
+
+  private LocalDateTime createdAt;
+
+
+  public void setCreatedAtNow() {
+    setCreatedAt(LocalDateTime.now());
+  }
```
위에서 기본값을 지정해주는건 어떨까요? (나중에 임의로 setCreatedAtNow 을 사용하려는 의도가 있다면 이대로 가도 좋을것 같네요)

f-lab-kyle commented on src/main/resources/mybatis/mapper/board/ArticleMapper.xml:
```diff
@@ -0,0 +1,102 @@
+  </insert>
+
+  <select id="selectMaxArticleId" resultType="int">
+    SELECT COALESCE(MAX(article_id) + 1, 1)
+    FROM bbs_articles
```
성능엔 괜찮을까요?

f-lab-kyle reviewed (commented):
좋아요 기능에서 트랜잭션의 문제가 될 경우는 없을까요? 만약 문제가 된다면 그것을 어떻게 처리하면 좋을지 고민해봐도 좋을것 같아요. (서버가 한대라면? 서버가 여러대라면?)

donsonioc2010 committed:
feat : 코드리뷰 적용

1. Article과 Status(추천,비추천)기능 분리
2. 추천,비추천 type에 대해서 Enum으로 분리
3. 추천,비추천 Enum분리에 따른 TypeHandler 구현

donsonioc2010 committed:
feat : Article Upsert 한번의 요청으로 이뤄지도록 수정
